privacyidea/lib/periodictask.py

```diff
@@ -0,0 +1,183 @@
+    :param ptask_id: ID of the database entry
+    :return: ID of the deleted entry
+    """
+    periodic_task = PeriodicTask.query.filter_by(id=ptask_id).first()
+    if periodic_task is None:
+        raise ParameterError("The periodic task with id {!r} does not exist".format(ptask_id))
+    return periodic_task.delete()
+
+
+def enable_periodic_task(ptask_id, enable=True):
+    """
+    Set the ``active`` flag of an existing periodic task to ``enable``.
+    If ``ptask_id`` refers to an unknown entry, a ParameterError is raised.
+    :param ptask_id: ID of the database entry
+    :param enable: New value of the ``active`` flag
+    :return: ID of the database entry
+    """
+    periodic_task = PeriodicTask.query.filter_by(id=ptask_id).first()
+    if periodic_task is None:
+        raise ParameterError("The periodic task with id {!r} does not exist".format(ptask_id))
+    periodic_task.active = enable
+    return periodic_task.save()
+
+
+def get_periodic_tasks(name=None, node=None, active=None):
+    """
+    Get a list of all periodic tasks, or of all tasks satisfying a filter criterion.
+    :param name: Name of the periodic task
+    :type name: unicode
+    :param node: Node for which periodic tasks should be collected. This only includes
+                 periodic tasks which are scheduled to run on ``node``.
+    :type node: unicode
+    :param active: This can be used to filter for active or inactive tasks only
+    :return: A (possibly empty) list of periodic task dictionaries
+    """
+    query = PeriodicTask.query
+    if name is not None:
+        query = query.filter_by(name=name)
+    if active is not None:
+        query = query.filter_by(active=active)
+    entries = query.all()
+    result = []
+    for entry in entries:
+        ptask = entry.get()
+        if node is None or node in ptask["nodes"]:
+            result.append(ptask)
+    return result
+
+
+def set_periodic_task_last_run(ptask_id, node, last_run_timestamp):
+    """
+    Write to the database the information that the specified
+    periodic task has been run on a node at a given time.
+    :param ptask_id: ID of the periodic task. Raises ParameterError if unknown.
+    :type ptask_id: int
+    :param node: Node name. It is not checked whether the task is scheduled to run on that node!
+    :type node: unioode
+    :param last_run_timestamp: Timestamp of the last run
+    :type last_run_timestamp: timezone-aware datetime object
+    """
+    periodic_task = PeriodicTask.query.filter_by(id=ptask_id).first()
+    if periodic_task is None:
+        raise ParameterError("The periodic task with id {!r} does not exist".format(ptask_id))
+    utc_last_run = last_run_timestamp.astimezone(tzutc()).replace(tzinfo=None)
+    periodic_task.set_last_run(node, utc_last_run)
+
+
+def get_scheduled_periodic_tasks(node, current_timestamp=None, interval_tzinfo=None):
```

Comment:
Is there a way of getting the list of available nodes?

Otherwise we might want to make the ``node`` parameter optional.

Reply:
Hah you're right, currently there is no way to obtain the available nodes. And I'm currently not sure how we would obtain such a list?

However, the ``get_scheduled_periodic_tasks`` function is intended to be called by the local cron runner to find out which tasks are scheduled to run on the local node. And at this point, the cron runner should know its local node name, and the other node names are not relevant then.

Reply:
So we leave it to the cron runner to determine the name of the node.

Can you please add something to the docstring like "This method is usually called by the local cron runner, who is aware of his local node name".

So the tasks are self contained and maybe we would add a node handling later.

I think for the UI it would be quite helpfull to have a lib function that returns the available node names so that the admin can choose accordingly and do not create a misspelling. (renaming nodes) Maybe we should put the node thingy into another issue!